sdk/storage/Azure.Storage.Queues/src/QueueClient.cs

```diff
@@ -2541,6 +2583,39 @@ protected internal virtual QueueClient WithClientSideEncryptionOptionsCore(Clien
                 }
             }
         }
+
+        private async Task<PeekedMessage[]> ToPeekedMessagesWithInvalidMessageHandling(
+            IEnumerable<PeekedMessageItem> peekedMessageItems,
+            bool async,
+            CancellationToken cancellationToken)
+        {
+            List<PeekedMessage> peekedMessages = new List<PeekedMessage>();
+
+            foreach (var peekedMessageItem in peekedMessageItems)
+            {
+                try
+                {
+                    peekedMessages.Add(PeekedMessage.ToPeekedMessage(peekedMessageItem, ClientConfiguration.MessageEncoding));
+                }
+                catch (FormatException) when (ClientConfiguration.InvalidMessageHandler != null)
+                {
+#pragma warning disable AZC0110 // DO NOT use await keyword in possibly synchronous scope.
+                    await ClientConfiguration.InvalidMessageHandler.RaiseAsync(
+                        new InvalidMessageEventArgs(
+                            queueClient: this,
+                            message: PeekedMessage.ToPeekedMessage(peekedMessageItem, QueueMessageEncoding.None),
+                            runSynchronously: !async,
+                            cancellationToken: cancellationToken),
+                        nameof(QueueClientOptions),
+                        nameof(QueueClientOptions.OnInvalidMessage),
+                        ClientConfiguration.ClientDiagnostics).ConfigureAwait(false);
```

Comment:
We should also pull this code into a `virtual async Task On[EventName]Async(QueueMesage receivedMessage, PeekedMessage peekedMessage, bool runSynchronously, CancellationToken cancellationToken)` method on QueueClient that folks can override.  This is useful for extensibility, mocking, and covered by https://learning.oreilly.com/library/view/framework-design-guidelines/9780135896457/ch05.xhtml#sec5_4.